go-controller/pkg/clustermanager/status_manager/status_manager.go

```diff
@@ -0,0 +1,156 @@
+	}
+}
+
+func (m *typedStatusManager[T]) Start() error {
+	_, err := m.informer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: m.updateStatus,
+		})
+	return err
+}
+
+type StatusManager struct {
+	abpRouteManager       *typedStatusManager[*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute]
+	egressFirewallManager *typedStatusManager[*egressfirewallapi.EgressFirewall]
+}
+
+func NewStatusManager(wf *factory.WatchFactory, ovnClient *util.OVNClusterManagerClientset) *StatusManager {
+	sm := &StatusManager{}
+	if config.OVNKubernetesFeature.EnableMultiExternalGateway {
+		sm.abpRouteManager = &typedStatusManager[*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute]{
+			informer: wf.APBRouteInformer().Informer(),
+			patch: func(ctx context.Context, namespace, name string, pt ktypes.PatchType, data []byte, opts metav1.PatchOptions,
+				subresources ...string) (*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute, error) {
+				return ovnClient.AdminPolicyRouteClient.K8sV1().AdminPolicyBasedExternalRoutes().Patch(ctx, name, pt, data, opts, subresources...)
+			},
+			updateStatusFunc: updateStatusAPBRoute,
+		}
+	}
+	if config.OVNKubernetesFeature.EnableEgressFirewall {
+		sm.egressFirewallManager = &typedStatusManager[*egressfirewallapi.EgressFirewall]{
+			informer: wf.EgressFirewallInformer().Informer(),
+			patch: func(ctx context.Context, namespace, name string, pt ktypes.PatchType, data []byte, opts metav1.PatchOptions,
+				subresources ...string) (*egressfirewallapi.EgressFirewall, error) {
+				return ovnClient.EgressFirewallClient.K8sV1().EgressFirewalls(namespace).Patch(ctx, name, pt, data, opts, subresources...)
+			},
+			updateStatusFunc: updateStatusEgressFirewall,
+		}
+	}
+	return sm
+}
+
+func (sm *StatusManager) Start() error {
+	klog.Infof("Starting StatusManager with params: abpRouteManager=%v, egressFirewallManager=%v",
+		sm.abpRouteManager != nil, sm.egressFirewallManager != nil)
+	if sm.abpRouteManager != nil {
+		err := sm.abpRouteManager.Start()
+		if err != nil {
+			return fmt.Errorf("failed to start abpRouteManager: %w", err)
+		}
+	}
+	if sm.egressFirewallManager != nil {
+		err := sm.egressFirewallManager.Start()
+		if err != nil {
+			return fmt.Errorf("failed to start abpRouteManager: %w", err)
+		}
+	}
+	return nil
+}
+
+// updateStatusAPBRoute returns if policy needs status update
+func updateStatusAPBRoute(route *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) ([]byte, bool) {
+	if route == nil || len(route.Status.Messages) == 0 {
+		return nil, false
+	}
+	newStatus := adminpolicybasedrouteapi.SuccessStatus
+	for _, message := range route.Status.Messages {
+		if strings.Contains(message, types.APBRouteErrorMsg) {
+			newStatus = adminpolicybasedrouteapi.FailStatus
+			break
+		}
+	}
+	return []byte(fmt.Sprintf(`{"apiVersion":"k8s.ovn.org/v1","kind":"AdminPolicyBasedExternalRoute","status":{"status":"%s"}}`, newStatus)),
```

Comment:
From what I know, not really. I tried marshalling objects, but in that case all fields of Message object will be added to the patch, but we don't really want that, since cluster manager only changes Status.Status and nodes only change Status.Messages and Status.LastTransitionTime. The problem is that nil/empty fields are treated by patch as a request to delete/reset the field, and even though you can't reset fields owned by other managers, the request doesn't really express what I want to happen. Another problem is that it breaks unit tests, since test apiserver is not smart enough to evaluate ownership conflicts, and cleans up unset fields :) 
Another way to do that is to create a special object to represent the status with desired json marshalling, but just using a string looks much simpler to me. lmk what you think

Reply:
ummm, isn't that a problem of our CRDs? We are not using pointers and `omitempty` as we should? Maybe we can change that now that we are at it if it is backward compatible.

Reply:
not really. because patching is changing a part of CRD, e.g. you may want to omit some required fields, just because you don't want to change them with a patch, while CRD will define that field as required and shouldn't be omitted.

Reply:
OK. But when we patch other things elsewhere I don't see us including the `{"apiVersion":"k8s.ovn.org/v1","kind":"AdminPolicyBasedExternalRoute"` part. Why would we need that here?

Reply:
ok, it wasn't easy, but I regenerated all clients :D
check last 2 commits to see how it looks, I left them separate to potentially backport only the first 2